File: Medium_Python/56_Merge_intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        #先对区间按开始时间进行排序，若下一个开始时间小于上一个结束时间，则合并两个区间，否则添加新的区间
        #注意合并时候的右边边界
        res=[]
        if len(intervals)==0:
            return res
        intervals.sort(key=lambda x: x[0])
        res.append(intervals[0])
        for i in range(1,len(intervals)):
            if intervals[i][0]>res[-1][1]:
                res.append(intervals[i])
            else:
                res[-1][1]=max(intervals[i][1],res[-1][1])
        return res
             

File: Medium/56_Merge_intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> res;
        sort(intervals.begin(),intervals.end());
        int fl = intervals[0][0];
        int fr = intervals[0][1];
        res.push_back({intervals[0][0],intervals[0][1]});
        for(int i = 1; i < intervals.size();i++)
        {
            int sl = intervals[i][0];
            int sr = intervals[i][1];
            
            if (sl-fr > 0)
            {
                res.push_back({intervals[i][0],intervals[i][1]});
            }
            else
            {
                int resback = res.back()[1];
                int resfront = res.back()[0];
                res.pop_back();
                res.push_back({
                    resfront,max(intervals[i][1],resback)});   
            }
            fl = res.back()[0];
            fr = res.back()[1];
            
        }
        return res;
    }
};
